app/models/ems_refresh/save_inventory.rb

```diff
@@ -181,6 +181,29 @@ def save_vms_inventory(ems, hashes, target = nil)
     end
   end
 
+  # Convert all mapped hashes into actual tags and associate them with the object.
+  # The hash[:tags] object should be an array of hashes, probably created by
+  # the ContainerLabelTagMapping.map_labels method. Each hash in the array should
+  # have the following basic structure:
+  #
+  # {:category_tag_id=>139, :entry_name=>"foo", :entry_description=>"bar"}
+  #
+  # For now, the object must be a VmOrTemplate object.
+  #
+  def save_tags_inventory(object, hash)
+    return if hash.blank?
+    return unless object.kind_of?(VmOrTemplate)
+
+    tags = hash[:tags].collect do |tag_hash|
+      ContainerLabelTagMapping.find_or_create_tag(tag_hash)
```

Comment:
right, we should rewrite the find_or_create_tag, so we don't do the n+1 queries there

so we should first process all the tags to pickup the unique combinations. Then load all existing in one query and create the missing.

If this could lead to huge number of AR object in memory, it would be nice to store only id then.

is the find_or_create_tag used in some refresh already?